[tt-train] SiLU backward kernel implementation

mdragulaTT opened the pull request:
### Problem description
The SiLU backward operation was previously using a high-level composite `ttnn` [implementation](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L919-L948). This approach resulted in creating performance slowdown in training workloads.

This PR replaces the composite implementation with a custom fused kernel that computes the entire SiLU backward operation in a single kernel launch improving overall training performance.

### What's changed
- **Added custom SiLU backward kernel** with fused compute operations in `tt-metal/tt-train/sources/ttml/metal/ops/silu_bw/`
- **Replaced composite ttnn implementation** with direct kernel calls in `tt-metal/tt-train/sources/ttml/ops/unary_ops.cpp`
- **Added comprehensive test coverage** in `tests/ops/silu_op_test.cpp` including:
  - Kernel vs reference implementation validation across different scenarios
  - Block size alignment patterns: Wt % block_size = 0, 1, 2, 3 (where block_size = 4)
  - Memory patterns: small tensors (C=8) to very large tensors (C=1M elements)
  - Realistic training shapes: NanoLlama-like dimensions (B=2, S=64, C=512)
  - Multiple batch and sequence configurations for transformer workloads
  - Unaligned channel dimensions (C % 32 != 0)
- **Integrated kernel into training pipeline** with backward compatibility maintained

### 3-packing vs 4-packing approaches
The SiLU backward operation can be computed using two different mathematical formulations, each requiring different numbers of intermediate circular buffer (CB) packs:

#### 4-packing approach:
Uses the standard formula: `grad * sigmoid_result * (1 + input * (1 - sigmoid_result))`
1. `sigmoid(x)` → PACK
2. `1 - sigmoid(x)` → PACK  
3. `(1 - sigmoid(x)) * x + 1` → PACK
4. `sigmoid(x) * ((1 - sigmoid(x)) * x + 1)` → PACK
5. `grad * result`

#### 3-packing approach:
Uses the equivalent formula: `grad * (-1) * (x + x * (-sigmoid) + 1) * (-sigmoid)`
1. `-sigmoid(x)` → PACK
2. `x + x*(-sigmoid(x)) + 1` (utilizing `mul_tiles` accumulation) → PACK
3. `(x + x*(-sigmoid(x)) + 1) * (-sigmoid(x))` with negation → PACK
4. `grad * result`

**Performance comparison results:**
Both approaches showed nearly identical performance improvements in NanoLlama training (see results below), so we selected the **4-pack implementation** due to **its more intuitive pipeline structure**. In fact, neither option demonstrated significant advantages in terms of precision, robustness, or speed over the other.

### Performance results
Mean step times across different configurations (5000 steps for NanoLlama):
- **Forward-only kernel**: 238.11 ms (baseline for forward pass)
- **Forward + backward 4-packs kernel**: 228.50 ms (**4.0% improvement** over forward-only)
- **Forward + backward 3-packs kernel**: 228.51 ms (**4.0% improvement** over forward-only)

**Key observations:**
- Both 3-packing and 4-packing approaches provide nearly identical performance improvements
- The backward kernel integration shows measurable training speedup
- Performance improvements are consistent across the entire training run (5000 steps)

The implementation maintains full numerical correctness while providing significant performance improvements for SiLU operations in the training pipeline.

Step time comparison across all configurations (5000 steps for NanoLlama):
<img width="2000" height="1000" alt="image" src="https://github.com/user-attachments/assets/aa6c4b0d-ce35-47c8-b0b1-85b8c48db9bc" />

Performance comparison (5000 steps for NanoLlama):
<img width="2000" height="1000" alt="image" src="https://github.com/user-attachments/assets/0ffd8df6-8354-4cbd-940c-23bb25f43420" />


Loss comparison showing numerical correctness (5000 steps for NanoLlama):
<img width="2000" height="1000" alt="image" src="https://github.com/user-attachments/assets/1d92cda8-1a3c-4415-a180-5defd1b1ae2a" />


### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

mdragulaTT commented:
## PR Updates:

### Changes Made

#### 1. Compute Kernel Cleanup
- Removed `working_register` variable from compute kernel as requested by Roman

#### 2. Reader Optimization
- Removed reading garbage in reader 

#### 3. Precision Comparison Analysis
As requested by Denys, I conducted a precision comparison with the following methodology:

##### Methodology
- **Baseline Implementation**: SiLU backward using `xt`
- **Test Configurations**: 
  - Composite implementation
  - 4 kernel setups: 3 packs vs 4 packs, using CBs for intermediate results with bfloat16 or float32
- **Evaluation Metric**: RMSE between baseline and each implementation
- **Test Approach**: 3 different tensor shapes, each tested with 10 runs using different random seeds
- **Statistical Analysis**: Mean values computed across all runs for robust results

##### Test Shapes
- **Small**: [1, 1, 1, 8] - 8 elements
- **NanoLlama-like**: [2, 1, 64, 512] - 65,536 elements
- **Large**: [1, 1, 1, 32768] - 32,768 elements

##### Key Findings

| Shape | Kernel (3p, bf16) | Kernel (3p, f32) | Kernel (4p, bf16) | Kernel (4p, f32) | Composite | **Winner** |
|-------|-------------------|------------------|-------------------|------------------|-----------|------------|
| Small [1,1,1,8] | 1.77e-03 | 2.73e-03 | 1.77e-03 | 2.74e-03 | **1.71e-03** | **Composite** |
| NanoLlama [2,1,64,512] | 2.55e-07 | 3.76e-07 | 2.55e-07 | 3.76e-07 | **2.13e-07** | **Composite** |
| Large [1,1,1,32768] | 5.10e-07 | 7.52e-07 | 5.10e-07 | 7.52e-07 | **4.25e-07** | **Composite** |

##### Surprising Results
1. **Composite Precision Advantage**: Composite implementation has slightly better precision across all shapes, but differences are small
2. **3-pack vs 4-pack**: Both kernel configurations show nearly identical precision performance
3. **Float32 Disappointment**: CBs using float32 do **not** provide better precision than bfloat16 variants :(
4. **Scale Consistency**: Precision patterns remain consistent across different tensor sizes

##### Statistical Confidence
- **10 runs per configuration**: Each result represents mean across 10 independent runs
- **Different random seeds**: Ensures statistical robustness (seeds 42-51)

##### Summary
- **Primary Goal**: Choose between 3-pack vs 4-pack kernel configurations for optimal precision
- **Kernel Performance**: Kernel implementations provide significant performance boost compared to composite
- **Precision Analysis**: While composite has slightly better precision, the differences are small and acceptable
- **Configuration Choice**: Both 3-pack and 4-pack show similar precision and performance, so either can be chosen
- **Surprising Finding**: Float32 format for intermediate data storage does **not** provide better precision than bfloat16 :(
- **Recommendation**: Use kernel implementation for performance benefits; precision differences are minimal and acceptable for training

##### Detailed Results by Shape
Complete precision metrics are available below, showing mean values across 10 runs for each configuration and shape.
```
Small Shape [1, 1, 1, 8] (8 elements) Backward Precision:
  Kernel (3 packs, bfloat16) - Mean Max Error: 3.51e-03, Mean Mean Error: 1.27e-03, Mean RMSE: 1.77e-03
  Kernel (3 packs, float32)  - Mean Max Error: 5.68e-03, Mean Mean Error: 1.85e-03, Mean RMSE: 2.73e-03
  Kernel (4 packs, bfloat16) - Mean Max Error: 3.51e-03, Mean Mean Error: 1.26e-03, Mean RMSE: 1.77e-03
  Kernel (4 packs, float32)  - Mean Max Error: 5.68e-03, Mean Mean Error: 1.86e-03, Mean RMSE: 2.74e-03
  Composite                  - Mean Max Error: 3.68e-03, Mean Mean Error: 1.17e-03, Mean RMSE: 1.71e-03
  Winner                     - Backward: Composite

NanoLlama-like Shape [2, 1, 64, 512] (65536 elements) Backward Precision:
  Kernel (3 packs, bfloat16) - Mean Max Error: 1.02e-06, Mean Mean Error: 1.65e-07, Mean RMSE: 2.55e-07
  Kernel (3 packs, float32)  - Mean Max Error: 1.21e-06, Mean Mean Error: 2.39e-07, Mean RMSE: 3.76e-07
  Kernel (4 packs, bfloat16) - Mean Max Error: 1.02e-06, Mean Mean Error: 1.64e-07, Mean RMSE: 2.55e-07
  Kernel (4 packs, float32)  - Mean Max Error: 1.21e-06, Mean Mean Error: 2.40e-07, Mean RMSE: 3.76e-07
  Composite                  - Mean Max Error: 8.39e-07, Mean Mean Error: 1.39e-07, Mean RMSE: 2.13e-07
  Winner                     - Backward: Composite

Large Shape [1, 1, 1, 32768] (32768 elements) Backward Precision:
  Kernel (3 packs, bfloat16) - Mean Max Error: 2.04e-06, Mean Mean Error: 3.29e-07, Mean RMSE: 5.10e-07
  Kernel (3 packs, float32)  - Mean Max Error: 2.42e-06, Mean Mean Error: 4.78e-07, Mean RMSE: 7.52e-07
  Kernel (4 packs, bfloat16) - Mean Max Error: 2.04e-06, Mean Mean Error: 3.29e-07, Mean RMSE: 5.10e-07
  Kernel (4 packs, float32)  - Mean Max Error: 2.42e-06, Mean Mean Error: 4.79e-07, Mean RMSE: 7.52e-07
  Composite                  - Mean Max Error: 1.67e-06, Mean Mean Error: 2.77e-07, Mean RMSE: 4.25e-07
  Winner                     - Backward: Composite
```

mdragulaTT committed:
Remove 3-pack implementation and use float32 format for SiLU backward kernel

- Removed 3-pack implementation as it provides no performance or precision benefits over 4-pack
- Standardized on bfloat16 data format for intermediate computations as it surprisingly shows slight advantage over float32
- Simplified precision comparison test to only compute necessary RMSE metrics
- Cleaned up compute kernel by removing unused 3-pack and float32 variants

mdragulaTT commented:
Hi everyone, PTAL at this PR. I've cleaned it up and decided to go with the 4-pack implementation as it's more intuitive, and I'm using `bfloat16` data format as it surprisingly shows slightly better precision than float32, although the difference is marginal

rfurko-tt reviewed (approved):
Overall looks good, LGTM

please make sure to clean up according to comments

mdragulaTT committed:
SiLU backward kernel implementation

Replace high-level composite ttnn backward implementation with custom fused kernel for improved performance. The kernel computes the entire SiLU backward operation in a single kernel launch, improving training step times.

Key improvements:
- Custom compute kernel with fused SiLU backward operations
- Block-wise interleaved reads/writes
- Improves training step times
- Maintains numerical correctness with existing composite implementation

The implementation supports various tensor configurations including unaligned dimensions and scales from small tensors to 1M+ elements, with comprehensive test coverage for production workloads.

mdragulaTT commented on tt-train/sources/ttml/metal/ops/silu_bw/device/kernels/dataflow/reader_silu_bw_interleaved_start_id.cpp:
```diff
@@ -0,0 +1,80 @@
+
+    constexpr uint16_t one = 0x00003F80;  // (bfloat16)1.0 -> uint16_t
+    constexpr uint16_t five = 0x40A0;     // (bfloat16)5.0 -> uint16_t
+    // Here might be called generate tile for one, but hope it's not needed because we fiugure this out in a different
```
Done.

mdragulaTT commented on tt-train/sources/ttml/metal/ops/silu_bw/device/silu_bw_device_operation_types.hpp:
```diff
@@ -0,0 +1,25 @@
+
+// Output tensor specs and tensors
+using spec_return_value_t = std::vector<ttnn::TensorSpec>;
+using tensor_return_value_t = std::vector<ttnn::Tensor>;
```
Done.

mdragulaTT committed:
Merge branch 'main' into mdragula/silu-bwd
